File: voice_control/bridge/clients/cs/llm_client.cs
// llm_client.cs
// Requires the NetMQ NuGet package.
// This is a client for the Python LLMServer.

using System;
using NetMQ;
using NetMQ.Sockets;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Threading;

public class LLMClient : IDisposable
{
    private RequestSocket _client;
    private readonly string _endpoint;
    private readonly string _authToken;
    private int _idCounter = 0;

    public LLMClient(string endpoint = "tcp://127.0.0.1:8000", string authToken = null)
    {
        _endpoint = endpoint;
        _authToken = authToken;
    }

    public void Connect()
    {
        _client = new RequestSocket();
        _client.Connect(_endpoint);
        Console.WriteLine($"[LLMClient] Connected to {_endpoint}");
    }

    public void Disconnect()
    {
        if (_client != null)
        {
            _client.Dispose();
            _client = null;
        }
        Console.WriteLine("[LLMClient] Disconnected");
    }

    public JObject Query(string content, string role = "user")
    {
        var parameters = new JObject
        {
            ["content"] = content,
            ["role"] = role
        };
        return _Request("query", parameters);
    }

    private JObject _Request(string method, JObject parameters)
    {
        _idCounter++;
        var request = new JObject
        {
            ["jsonrpc"] = "2.0",
            ["method"] = method,
            ["params"] = parameters,
            ["id"] = _idCounter
        };

        if (!string.IsNullOrEmpty(_authToken))
        {
            request["auth_token"] = _authToken;
        }

        _client.SendFrame(request.ToString(Formatting.None));

        if (_client.TryReceiveFrameString(TimeSpan.FromSeconds(5), out var responseJson))
        {
            var response = JObject.Parse(responseJson);
            if (response["error"] != null)
            {
                throw new Exception($"RPC Error: {response["error"]}");
            }
            return response["result"] as JObject;
        }
        else
        {
            throw new TimeoutException("No response from server within the timeout period.");
        }
    }

    public void Dispose()
    {
        Disconnect();
    }

    // Example Usage
    public static void Main(string[] args)
    {
        using (var client = new LLMClient())
        {
            client.Connect();
            try
            {
                var response = client.Query("Hello from C#!");
                Console.WriteLine("Response from LLM: " + response);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }
    }
}

File: voice_control/bridge/clients/js/llm_client.js
// llm_client.js
// Requires the 'zeromq' npm package.
// This is a client for the Python LLMServer.

const zmq = require("zeromq");

class LLMClient {
    constructor(endpoint = "tcp://0.0.0.0:8000", authToken = null) {
        this.endpoint = endpoint;
        this.authToken = authToken;
        this.socket = new zmq.Request();
        this._idCounter = 0;
    }

    async connect() {
        this.socket.connect(this.endpoint);
        console.log(`[LLMClient] Connected to ${this.endpoint}`);
    }

    async disconnect() {
        this.socket.close();
        console.log("[LLMClient] Disconnected");
    }

    async query(content, role = "user") {
        const params = { content, role };
        return this._request("query", params);
    }

    async _request(method, params) {
        this._idCounter++;
        const request = {
            jsonrpc: "2.0",
            method: method,
            params: params,
            id: this._idCounter,
        };

        if (this.authToken) {
            request.auth_token = this.authToken;
        }

        await this.socket.send(JSON.stringify(request));
        const [result] = await this.socket.receive();
        const response = JSON.parse(result.toString());

        if (response.error) {
            throw new Error(`RPC Error: ${JSON.stringify(response.error)}`);
        }
        return response.result;
    }
}

// Example Usage
async function main() {
    const client = new LLMClient();
    await client.connect();
    try {
        const response = await client.query("Hello from Node.js!");
        console.log("Response from LLM:", response);
    } catch (e) {
        console.error("Error:", e.message);
    } finally {
        await client.disconnect();
    }
}

if (require.main === module) {
    main();
}

module.exports = LLMClient;